### ISSUE TITLE
[bug] Packages being marked as dirty on upload, when they are not dirty

### ISSUE THREAD

**Issue by Daniel-Roberts-Bose** [conan-io/conan#6330]
I have a weird bug here that I am fairly sure is related to the recent changes to enable parallel uploads.

I confirmed that rolling back to conan version 1.20.5, before this feature was implemented, fixed my issues completely.

In the steps to reproduce below, what is happening is that on the very last upload command, my packages are being marked as dirty and conan is deleting them, meaning they dont get uploaded. I believe this is related to the parallel changes because I am seeing logs come out of order, suggesting a threading issue. It seems like what is happening is that its running the follow line of code

https://github.com/conan-io/conan/blob/a872fad79ac2ae2df282ba1f81b17fead726b820/conans/client/cmd/uploader.py#L568

Before running this line of code:

https://github.com/conan-io/conan/blob/a872fad79ac2ae2df282ba1f81b17fead726b820/conans/client/cmd/uploader.py#L401

This is causing the package to be marked as dirty here:

https://github.com/conan-io/conan/blob/a872fad79ac2ae2df282ba1f81b17fead726b820/conans/util/files.py#L62

My suspicion is this yield call. I think perhaps the yield call is causing some weirdness and perhaps its going and executing another upload command before the first one is finished? I am unsure completely how the yield call works under the hood in python but it seems suspect to me without correct mutex protection

### Environment Details (include every applicable attribute)
  * Operating System+version: Linux, Ubuntu 16
  * Conan version: 1.22.0-dev
  * Python version: 3.6

### Steps to reproduce (Include if Applicable)
Create a package, this also does some intermediate rebuilds in the middle of the dependency graph in my case
`conan create . test/test` 

Compress one of those intermediate packages but dont upload yet
`conan upload PkgA/1.0.0@test/test -c --check --skip-upload`

Upload that package after checks have been performed and its been compressed
`conan upload PkgA/1.0.0@test/test -c --check`

### Logs (Executed commands with output) (Include/Attach if Applicable)
```
[2020-01-08T22:13:01.902Z] + conan upload PkgA/1.0.0@test/test:e52c9eb724d0973fa676c1bc641288614396a401 -r R1 -c
[2020-01-08T22:13:02.209Z] Uploading to remote 'R1':
[2020-01-08T22:13:02.209Z] 
Uploading PkgA/1.0.0@test/test to remote 'R1'    
[2020-01-08T22:13:04.354Z] Recipe is up to date, upload skipped
[2020-01-08T22:13:04.354Z] 
Uploading package 1/1: e52c9eb724d0973fa676c1bc641288614396a401 to 'R1'           
[2020-01-08T22:13:04.354Z] Compressing package...
[2020-01-08T22:13:37.289Z] WARN: conan upload PkgA/1.0.0@test/test:e52c9eb724d0973fa676c1bc641288614396a401: Removing conan_package.tgz, marked as dirty
```

Also, this may be related to https://github.com/conan-io/conan/issues/6141 but I am unsure. I will keep digging and upload anything else I find out. Let me know if you need any more information to help debug

**Comment by memsharded** [comment]
Hi @Daniel-Roberts-Bose 

I have tried to reproduce this, so far without success.
I might be missing some detail. For example, how is it possible that the  package is being uploaded if ``--all`` is not provided in the command line?

Maybe you can repeat the steps above, but before and after each command, you can ``ls`` or have a look to the conan cache. You should see (or miss) a file with the ``.dirty`` extension, that is called ``conan_package.tgz.dirty`` (you might need ``--all`` to all commands above to see the ``conan_package.tgz`` created), in the ``pkg/version/user/channel/package/ID/`` folder. 
Please let me know how it goes, thanks!

**Comment by Daniel-Roberts-Bose** [comment]
Hi @memsharded 

I will get back to you on running the ls commands between each operation but to answer your other question about how it's uploading without --all, we are specifying an exact package ID, I forgot to add that above. The actual reference we are passing to the upload command looks something like `PkgA/1.0.0@test/test:abc12345678foobar`

A little but of background also, I am not sure how relevant any of this will be but perhaps it will help. We are running a create command on a dependency that is very high up in the tree, meaning underneath it there are about 10 layers of dependencies beneath it. Naturally, the graph sees that some package IDs are missing and builds those IDs during the install step (`--build=missing`) at the very end we then take the output of the `create.json` and parse it for package IDs to upload. We then have a list of IDs and loop through and first compress them all, then loop through again and upload them all. Also note here, that each pkg often has multiple different package IDs to be uploaded because we also are running multiple creates with different settings and options

I'm not sure if any of that really helps to debug an upload problem but there is some background :)

**Comment by memsharded** [comment]
I have tried using the complete reference ``PkgA/1.0.0@test/test:abc12345678foobar`` and still not able to reproduce.

I don't think the fact that it is a dependency deep in the graph should affect this. The process of creating the ``conan_package.tgz`` and uploading it, is totally decoupled of the package creation process, so it should be exactly the same for any package. It might be something in that package recipe that could be different?

So another question: Is the upload being executed concurrently? Is it possible that same package is being concurrently zipped (upload --skip-download, two references in the create.json that are the same, e.g. due to some build-requires)?

If you are able to try the above step by step, and checking that ``.dirty`` file in the cache, it would be useful also to get the terminal output of that package that is dirty. Thanks!

**Comment by Daniel-Roberts-Bose** [comment]
The build is not being executed concurrently. Each upload is done in serial and not parallel. At least, we aren't intentionally doing things in parallel unless there is a bug here somewhere. I checked the create.json and it doesn't look like there are two references there for the same package.

I am trying to above steps now to check for the dirty file. The build that reliably reproduces this takes a while so I will get back to you later today :)

**Comment by memsharded** [comment]
> Perhaps a question that may answer this whole problem is, does the conan upload command compress things before exiting? Or does the command return and do things in the background somehow? I have always assumed and from my experience it will compress things before returning but we are running many upload commands in a loop, one after the other.

No, there is nothing like a background task, everything should be completely serial. No problem, take your time and let me know about your findings, thanks!

**Comment by Daniel-Roberts-Bose** [comment]
I ran some testing and was not able to reproduce it again locally for some reason. I managed to force the bug to appear by running two upload commands on the same package at the same time. So next steps are to see if the reason there are two upload commands running, is due to the way we are calling the conan command or due to some internal conan defect. 

I will continue to test and get back to you

**Comment by memsharded** [comment]
Thanks @Daniel-Roberts-Bose for trying to reproduce. I have been having a look at the code too, can't find any place that could originate this dirty either, besides concurrent uploads by different processes. So looking forward your findings in your CI jobs, thanks!

**Comment by KerstinKeller** [comment]
Hi, I am encountering the exact same thing in our CI, where we get a `Removing conan_package.tgz, marked as dirty`
We have different stages, and one is the actual upload. Iterating over all recipes we're doing something like:

```
conan upload package/version@user/channel --all --confirm -r artifactory-test
```

This is an excerpt from our Jenkins log:
```
14:39:45  D:\...>python build.py upload --packages=qt* 
14:39:46  Uploading to remote 'artifactory-test':
14:39:46  
Uploading qt/5.11.1@user/stable to remote 'artifactory-test'                            
14:39:46  Recipe is up to date, upload skipped
14:39:46  
Uploading package 1/2: 964cba37adb5d2352f2e8e7a977f4b69d69c2e61 to 'artifactory-test'    
14:39:47  Compressing package...
14:40:19  
Uploading package 2/2: caa445deac6f2f357801d240079788bf3f962b1f to 'artifactory-test'    
14:40:19  Compressing package...
14:40:58  WARN: qt/5.11.1@user/stable:964cba37adb5d2352f2e8e7a977f4b69d69c2e61: Removing conan_package.tgz, marked as dirty
14:40:58  WARN: qt/5.11.1@user/stable:caa445deac6f2f357801d240079788bf3f962b1f: Removing conan_package.tgz, marked as dirty
```

I do not have the slightes clue why they could be marked as dirty. Nothing happens concurrently.
The actual conan command is being called by a Python script.

We have enabled revisions, and there is a already a package for the same hash, but an older revision on the server. I don't know if this relates to the problem at all.

One thing I realized though: conan_package.tgz is created inside the package folder. Could the mere existance of conan_package.tgz cause the package to be dirty? But then I guess that would apply to all packages...

![image](https://user-images.githubusercontent.com/13848742/73943985-116f1700-48f2-11ea-93b4-751a07cadf7f.png)

@Daniel-Roberts-Bose Have you been able to solve this?

**Comment by Daniel-Roberts-Bose** [comment]
@KerstinKeller are you using conan 1.22? it was released earlier this week so I am happy other people are now seeing this issue and it wasnt just me :) 

I have no found a work around or a fix, or even what the root cause was. It was such a blocker for us that I had to revert to using conan version 1.20.5, so I haven't gotten any additional testing on 1.22

My initial suspicion still stands that there may be something lurking in the `--parallel` upload option because I added in logging to the conan library and noticed it was doing upload function calls out of order

**Comment by KerstinKeller** [comment]
Yes, I am using Conan 1.22.
I just reverted on 1.20.5, and can confirm that the upload works as expected:

```
16:01:59  D:\...>python build.py upload --packages=qt* 
16:02:02  Uploading to remote 'artifactory-test':
16:02:02  Uploading qt/5.11.1@user/stable to remote 'artifactory-test'
16:02:02  Recipe is up to date, upload skipped
16:02:02  Uploading package 1/2: 964cba37adb5d2352f2e8e7a977f4b69d69c2e61 to 'artifactory-test'
16:02:05  Compressing package...
16:03:33  [==================================================]
16:03:33  
16:03:33  
16:04:05  
Uploading conan_package.tgz                                           
16:04:05  
16:04:05  
Uploading conaninfo.txt                                               
16:04:05  
16:04:05  
Uploading conanmanifest.txt                                           
16:04:05  
16:04:05  Uploading package 2/2: caa445deac6f2f357801d240079788bf3f962b1f to 'artifactory-test'
16:04:09  Compressing package...
16:06:09  [==================================================]
16:06:09  
16:06:09  
16:06:43  
Uploading conan_package.tgz                                           
16:06:43  
16:06:43  
Uploading conaninfo.txt                                               
16:06:43  
16:06:43  
Uploading conanmanifest.txt                                           
16:06:43  
16:06:45  WARN: qt/5.11.1@user/stable:964cba37adb5d2352f2e8e7a977f4b69d69c2e61: Removing conan_package.tgz, marked as dirty
16:06:45  WARN: qt/5.11.1@user/stable:caa445deac6f2f357801d240079788bf3f962b1f: Removing conan_package.tgz, marked as dirty
```

So the upload is done anyways, though I still get the warning.

@memsharded what do you think can cause the dirty tag?
What else can I do to help you understand the problem?

**Comment by memsharded** [comment]
This is weird. The only place in code where that conan_package.tgz.dirty is created is:

```python
def compress_files(files, symlinks, name, dest_dir, output=None):
    t1 = time.time()
    # FIXME, better write to disk sequentially and not keep tgz contents in memory
    tgz_path = os.path.join(dest_dir, name)
    with set_dirty_context_manager(tgz_path), open(tgz_path, "wb") as tgz_handle:
        tgz = gzopen_without_timestamps(name, mode="w", fileobj=tgz_handle)
```
which happens while uploading, previous to the upload. This only remains dirty if the .tgz is not created successfully.

It is also not clear why the ``16:06:45  WARN: qt/5.11.1@user/stable:96...e61: Removing conan_package.tgz, marked as dirty`` appears AFTER the upload. That is done before the actual transfer, that appears sooner.

**Comment by KerstinKeller** [comment]
Just to be clear, what does dirty mean here? That the package folder was modified between being created an being compressed?
So for the Jenkins job, it's just `build`, and then some time later `upload`. I don't know who would, or could touch the package directory in the meantime.

The output where the warning appears AFTER the upload, that with Conan 1.20.5. With Conan 1.22.0, there is the warning, no upload, but also no error message.

**Comment by memsharded** [comment]
> Just to be clear, what does dirty mean here? That the package folder was modified between being created an being compressed?

Dirty is a flag to make sure the process of creating the .tgz finished succesfully. If something breaks while zipping the .tgz, we want to make sure that such .tgz, which is incomplete and broken, is not upload nor used. It is not related to the usage of the package, only about zipping it to upload it zipped.

**Comment by Daniel-Roberts-Bose** [comment]
From investigation when i was looking at this, there are two levels of a package being dirty:
1. If there are untracked or modified files in the source code somewhere, during build
2. If during the packging process there are modifed files somewhere

I think the issue was number 2 here, where during package and upload it was being marked dirty via a `.dirty` file being added to the package

**Comment by KerstinKeller** [comment]
Ah ok, so when this `.dirty` file is there, it suggest that something went wrong compressing the files?
Or could it be that the '.dirty' file could not be removed?

I am playing around a bit, and what I did notice is, that the upload failed on a package what is quite big (~150 MB) and the compression takes quite some time, but with some ssmaller packages I did not encounter any problems just now.

Has anything regarding compression changed? Are there timeouts?  I will investigate more and post anything which might be relevant.

**Comment by memsharded** [comment]
Some issues related to upload from Conan 1.21 are been fixed in https://github.com/conan-io/conan/pull/6486/files. If possible to run from sources from that branch, that would be very useful info too, thanks!

**Comment by KerstinKeller** [comment]
Ok, so I think the two issues are actually unrelated.

I can confirm that the `conan_package.tgz` is actually corrupt when the message is displayed. So it does not have the size it should have (~128MB), but it's only 50MB in size and cannot be opened with 7zip.

So I guess that the warning is correct (though somewhat missleading). Instead it should be warned (or acknowledged) that the packaging process has failed (for whatever reason).
Also, though it says `Removing conan_package.tgz, marked as dirty`, nothing is removed here.

Now the really interesting thing: one clean machines it always runs. But on machines where the job has run before, it does fail to compress the `conan_package.tgz` file.
In all jobs, a `conan remove '*' -f` is executed to have a clean cache. But even though, I would expect that before copying to a `package/<rev>` folder, conan would clean that folder.

So it seems very unintuitive to me, why a job would fail to compress the data on a build agent where that package has been build before (successfully), but succeed in compressing in case of a `clean` build agent.

This was all observed with Conan 1.22.0, but now I will try with the branch from the PR.

**Comment by KerstinKeller** [comment]
The output to running conan from the branch has been reported in #6482 .
But they do seem unrelated.

What is still strange is that even though Conan warns about the corrupt `conan_package.tgz` file, is that an upload is still attempted and the file(s) don't get deleted, altough that should happen, according to the code.

```
if is_dirty(tgz_path):
  self._output.warn("%s: Removing %s, marked as dirty" % (str(ref), f))
  os.remove(tgz_path)
  clean_dirty(tgz_path)
```

**One more quick question:**
The timestamps suggests that the actual `qt/5.11.1/user/stable/package/<rev>` folders have not been deleted when executing `conan remove '*' -f`.
What is the correct command to clean the whole local cache, properly and entirely?

**Comment by memsharded** [comment]
Hi @KerstinKeller 

I think the main issue was that both the zipping of the ``conan_package.tgz`` and the transfers, when they are for large packages that take a lot of time (+30secs) it did crash. Leaving the .dirty state, skipping the upload without reporting it.

I strongly believe that https://github.com/conan-io/conan/pull/6486 will fix it. I will wait for @KerstinKeller confirmation, but meanwhile merging this to 1.22.1, to pass CI and have the release ready.

For the other questions, I think that deserves a new issue, might be something unrelated (sometimes, the build scripts might mark some files as Read-Only, and then that is not removed by ``conan remove * -f``, which should clean all packages). To completely clean the cache, it can be ``rm -rf /home/user/.conan`` or path to the cache, followed by a ``conan config install`` assuming that the cache is configured that way. In any case, it is worth opening a new issue to track it.

**Comment by KerstinKeller** [comment]
Hi @memsharded,

so from my side, the upload looks good. Compresssion works, and everything gets uploaded that should get uploaded.

However, I still get that warning, but somewhat unjustified? `conan_package.tgz` is present and non-corruped, and there is no `.dirty` file in the packages folder.
But maybe this is related to the non-clean cache in the first place.

```
08:40:13  D:\...>python build.py upload --packages=qt* 
08:40:16  Uploading to remote 'artifactory-test':
08:40:16  Uploading qt/5.11.1@user/stable to remote 'artifactory-test'                            
08:40:16  Recipe is up to date, upload skipped
08:40:16  Uploading package 1/2: 964cba37adb5d2352f2e8e7a977f4b69d69c2e61 to 'artifactory-test'    
08:40:20  Compressing package...
08:42:47  ....Uploading conan_package.tgz -> qt/5.11.1@user/stable:964c
08:42:47  Uploading conaninfo.txt -> qt/5.11.1@user/stable:964c
08:42:47  Uploading conanmanifest.txt -> qt/5.11.1@user/stable:964c
08:42:47  Uploading package 2/2: caa445deac6f2f357801d240079788bf3f962b1f to 'artifactory-test'    
08:42:49  Compressing package...
08:44:58  ...Package is up to date, upload skipped
08:45:00  WARN: qt/5.11.1@user/stable:964cba37adb5d2352f2e8e7a977f4b69d69c2e61: Removing conan_package.tgz, marked as dirty
08:45:00  WARN: qt/5.11.1@user/stable:caa445deac6f2f357801d240079788bf3f962b1f: Removing conan_package.tgz, marked as dirty
```

**Comment by memsharded** [comment]
Hi @KerstinKeller 

Thanks very much for the feedback. I keep investigating the ``conan_package.tgz`` dirty status. Is it possible that you are also using ``conan export-pkg`` in your flow? We have fixed an issue there for this 1.22.1 release in https://github.com/conan-io/conan/pull/6498

Do you have ``short_paths=True`` in those recipes? There isn't a ``conan_package.tgz.dirty`` file in the cache, not before, not after your command?

**Comment by KerstinKeller** [comment]
Hi @memsharded,

I tested the `release/1.22.1` branch, since your changes were already merged there.

So when I ran the test with the log posted, the package folders where there from the previous Jenkins run and contained a `.dirty` file before the run. After the run, that file was gone, but the warning was in the output.

Running the job a few more times, the warning does NOT occur any more.

So yes, there was a `.dirty` file before the command, but not after. This might still trigger the warning.

We are not using `short_paths`. We are not using `conan export-pkg`, but we are running first `conan remove * -f`, then exporting everything with `conan export ...`, building with `conan create ... -b outdated` and finally uploading with `conan upload ... --all --confirm`.

**Comment by memsharded** [comment]
> I tested the release/1.22.1 branch, since your changes were already merged there.

Great, perfect

> So when I ran the test with the log posted, the package folders where there from the previous Jenkins run and contained a .dirty file before the run. After the run, that file was gone, but the warning was in the output. Running the job a few more times, the warning does NOT occur any more.
> So yes, there was a .dirty file before the command, but not after. This might still trigger the warning.

Yes, then everything seems to make sense. The previous bug could cause the .dirty to be there, but once it is cleaned the first time that (without the bug) the upload process finish successfully, the .dirty is not there anymore. And the warning did make sense.

What doesn't make sense at all is that the warning shows AFTER the package is uploaded. I am trying to follow every possible execution trace, parallel or not, and I cannot find any scenario in which such warning is done after all the other upload messages. I keep investigating this, I would like to make sure this is not hiding any other possible issue.

**Comment by memsharded** [comment]
Finally identified the reason of the WARN about conan_package.tgz being dirty in such "random" order.
The thing is that the WARN messages are routed to the ``stream_err`` stream, which would be ``sys.stderr``, and then the calling process can get it in different not strictly sequential order.

So it seems a non-issue. I will create a new issue to discuss if we could have a fully deterministic output in Conan 2.0 (this might be breaking if someone is using ``sys.stderr``)

Thanks again very much for all the feedback. Lets release 1.22.1 asap.

**Comment by gsemet** [comment]
We encounter this issue when the path to .dirty file is too long. The only solution is to enable ShortPath.
